### PR TITLE
Added context support

### PR DESCRIPTION
**Pull request by lelandrichardson**
to: @ljharb @goatslacker 

Adds support for context with `mount` and `shallow`.

The API looks like:

```jsx
const context = { name: 'foo' };
const wrapper = mount(<Component />, { context });
```

```jsx
const context = { name: 'foo' };
const wrapper = shallow(<Component />, { context });
```

Additionally, there is a `.setContext(newContext)` prototype method on each wrapper class.

Fixes https://github.com/airbnb/enzyme/issues/49

Note: this will constitute a minor change (new feature, backwards compatible)

### CODE REVIEW

File: docs/api/ReactWrapper/setContext.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+expect(wrapper.text()).to.equal('foo');
+wrapper.setContext({ name: 'bar' });
+expect(wrapper.text()).to.equal('bar');
+wrapper.setContext({ name: 'baz' });
+expect(wrapper.text()).to.equal('baz');
+```
+
+#### Common Gotchas
+
+- `.setContext()` can only be used on a wrapper that was initially created with a call to `mount()`
```
Comment (ljharb) [inline docs/api/ReactWrapper/setContext.md]: Is this guaranteed, ie, will it throw an exception? I don't see any tests that cover this case.

File: docs/api/ShallowWrapper/setContext.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+expect(wrapper.text()).to.equal('foo');
+wrapper.setContext({ name: 'bar' });
+expect(wrapper.text()).to.equal('bar');
+wrapper.setContext({ name: 'baz' });
+expect(wrapper.text()).to.equal('baz');
+```
+
+#### Common Gotchas
+
+- `.setContext()` can only be used on a wrapper that was initially created with a call to `mount()`
```
Comment (ljharb) [inline docs/api/ShallowWrapper/setContext.md]: the example immediately above has a wrapper initially created with a call to shallow, not mount

File: src/ReactWrapperComponent.jsx [hunk 3]
```diff
@@ -8,45 +8,75 @@ import React, { PropTypes } from 'react';
...
+
+    getInitialState() {
+      return {
+        props: this.props.props,
+        context: this.props.context,
+      };
+    },
+
+    setChildProps(props) {
+      return new Promise(resolve => this.setState({ props }, resolve));
```
Comment (ljharb) [inline src/ReactWrapperComponent.jsx]: not part of this PR, but why doesn't `this.setState` return a promise rather than taking a callback? is that to mirror react's API?

File: docs/api/ReactWrapper/setContext.md (hunk 1 above)
Comment (lelandrichardson) [inline docs/api/ReactWrapper/setContext.md]: good call. will add some tests.

File: docs/api/ShallowWrapper/setContext.md (hunk 2 above)
Comment (lelandrichardson) [inline docs/api/ShallowWrapper/setContext.md]: good catch.

File: src/ReactWrapperComponent.jsx (hunk 3 above)
Comment (lelandrichardson) [inline src/ReactWrapperComponent.jsx]: because `this.setState()` can or maybe will be an asynchronous call... but it doesn't seem to be in react 0.13 or 0.14...  originally I had set up the API in such a way that it returned a promise, but it doesn't appear to be needed in most cases (and is thus not used right now). This promise code could be removed (maybe separate PR)

Comment (ljharb) [inline src/ReactWrapperComponent.jsx]: I like everything returning a Promise :-) it's future-compatible with async/await!

**Comment by lelandrichardson** [comment]
@ljharb tests added. comments addressed.

**Comment by ljharb** [comment]
Changes LGTM - @goatslacker should weigh in on the context stuff tho since he's got more experience with it than I

**Comment by lelandrichardson** [comment]
Looking at the coverage info with the line numbers mismatched is a PITA. I'm going to submit a followup PR that fixes istanbul line numbers, as well as improve coverage where needed

File: src/ReactWrapperComponent.jsx [hunk 4]
```diff
@@ -8,45 +8,75 @@ import React, { PropTypes } from 'react';
...
-    if (inst === null) {
-      return component;
-    }
-    return inst;
+  if (options.context && node.type.contextTypes) {
+    // For full rendering, we are using this wrapper component to provide context if it is
+    // specified in both the options AND the child component defines `contextTypes` statically.
+    // In that case, we define both a `getChildContext()` function and a `childContextTypes` prop.
+    Object.assign(spec, {
+      childContextTypes: node.type.contextTypes,
```
Comment (goatslacker) [inline src/ReactWrapperComponent.jsx]: You're mutating spec here. Is that ok? Seems like a bad idea

Comment (goatslacker) [inline src/ReactWrapperComponent.jsx]: Yeah it's ok.

**Comment by goatslacker** [comment]
LGTM

File: src/ReactWrapper.js [hunk 5]
```diff
@@ -54,10 +54,12 @@ export default class ReactWrapper {
     }
 
     if (!root) {
+      const ReactWrapperComponent = createWrapperComponent(nodes, options);
       this.component = renderIntoDocument(
         <ReactWrapperComponent
           Component={nodes.type}
           props={nodes.props}
+          context={options.context}
           />
```
Comment (blainekasten) [inline src/ReactWrapper.js]: not due to this diff, but the airbnb styleguide shows this closing wrapper to be back 2 spaces. https://github.com/airbnb/javascript/tree/master/react#alignment

Comment (goatslacker) [inline src/ReactWrapper.js]: tru dat

**Comment by kujon** [comment]
Super useful PR. May I ask what is the reason for `render` not having this functionality?
